jose/jwk.py

```diff
@@ -245,7 +259,191 @@ def verify(self, msg, sig):
+            # PEM wrapper. This may take a bit of guesswork...
+            der = self.pem_to_der(key)
+            # The key dictates the curve, this emulates the ecdsa lib
+            (self.curve, raw_key, raw_pub) = self.asn_to_raw(der, self.curve)
+            self.prepared_key = pyelliptic.ECC(
+                curve=self.curve,
+                privkey=raw_key,
+                pubkey=raw_pub,
+                hasher=sha_map[algorithm])
+            return
+        raise JWKError('Unable to parse an ECKey from key: %s' % key)
+
+    def repad(self, st):
+        """Add base64 padding back to the end of a stripped character
+        sequence
+
+        """
+        pad = '====' if isinstance(st, six.text_type) else b'===='
+        return st + pad[len(st) % 4:]
+
+    def pem_to_der(self, pem):
+        lines = pem.strip().split(b"\n")
+        return b''.join([line.strip() for line in lines if b'---' not in line])
+
+    def bitstring_to_str(self, bitstring):
+        """Convert an ASN1 BitString to a character array."""
+        if isinstance(bitstring, univ.OctetString):
+            return bitstring.asOctets()
+        if isinstance(bitstring, univ.BitString):
+            # Convert using a 2.7 safe method.
+            lh = hex(int(
+                ''.join(map(str, bitstring)), base=2))[2:].replace('L', '')
+            return unhexlify('0' * (len(lh) % 2) + lh)
+
+    def asn_to_raw(self, candidate, curve):
+        """Extract the ASN1 information and return the curve and key pairs."""
+        decoded = base64.urlsafe_b64decode(self.repad(candidate))
+
+        # if it's already raw... (Most likely a public key)
+        if len(decoded) == 64:
+            return curve, None, "\04" + decoded
+        if decoded[0] == "\04":
+            return curve, None, decoded
+
+        try:
+            asn_set = decoder.decode(decoded)[0]
+        except:
+            raise JWKError("Invalid EC Key")
+        pri_key = self.bitstring_to_str(asn_set[1])
+        pub_key = None
+        # A private key starts with a Integer(1)
+        if (isinstance(asn_set[0], univ.Integer) and
+                asn_set[0] == 1):
+            # Followed by the OID
+            curve = self.curve_oids.get(asn_set[2])
+            if curve:
+                # And finally the public key
+                pub_key = self.bitstring_to_str(asn_set[3])
+        # A public key starts with a sequence
+        if isinstance(asn_set[0], univ.Sequence):
+            # confirm that the public key curve matches up
+            # with the OID pair type (Also includes the public key OID,
+            # which we ignore)
+            pcurve = self.curve_oids.get(asn_set[0][1])
+            if pcurve:
+                curve = pcurve
+            # Shuffle the key to the right return.
+            pub_key = pri_key
+            pri_key = None
+        if not curve:
+            raise JWKError("Invalid EC curve type key specified.")
```

Comment:
Nit: It's not really that it was specified wrong, it's that the supplied data isn't one of our known formats.